lib/primitive/lib/Cardano/Wallet/Primitive/Types/Coin.hs

```diff
@@ -94,16 +106,11 @@ newtype Coin = Coin
     { unCoin :: Natural
     }
     deriving stock (Ord, Eq, Generic)
-    deriving (Read, Show) via (Quiet Coin)
-
--- | The 'Semigroup' instance for 'Coin' corresponds to ordinary addition.
---
-instance Semigroup Coin where
-    -- Natural doesn't have a default Semigroup instance.
-    (<>) = add
-
-instance Monoid Coin where
-    mempty = Coin 0
+    deriving (Read, Show) via Quiet Coin
+    deriving (Commutative, Semigroup, Monoid, MonoidNull) via Sum Natural
+    deriving (LeftReductive, RightReductive, Reductive) via Sum Natural
+    deriving (LeftGCDMonoid, RightGCDMonoid, GCDMonoid) via Sum Natural
+    deriving (OverlappingGCDMonoid, Monus) via Sum Natural
```

Comment:
looks serious :-)

Reply:
For this PR, the classes we care about are primarily:
- [`MonoidNull`](https://hackage.haskell.org/package/monoid-subclasses-1.2.3/docs/Data-Monoid-Null.html): provides the `null` operation: `null x == (x == mempty)`.
- [`Reductive`](https://hackage.haskell.org/package/monoid-subclasses-1.2.3/docs/Data-Semigroup-Cancellative.html#t:Reductive): provides the reductive subtraction operation: `</> :: m -> m -> Maybe m`
- [`Monus`](https://hackage.haskell.org/package/monoid-subclasses-1.2.3/docs/Data-Monoid-Monus.html#t:Monus): provides the [monus](https://en.wikipedia.org/wiki/Monus) (truncated) subtraction operation: `<\> :: m -> m -> m`

The other classes are superclasses of the classes we're interested in, which means we must derive them too. (It would be nice if there was a way to auto-derive superclasses without explicitly listing them, but there isn't.)